x-pack/elastic-agent/pkg/agent/application/upgrade/rollback.go

```diff
@@ -0,0 +1,158 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package upgrade
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path/filepath"
+	"runtime"
+	"strings"
+	"time"
+
+	"github.com/hashicorp/go-multierror"
+
+	"github.com/elastic/beats/v7/libbeat/common/backoff"
+	"github.com/elastic/beats/v7/x-pack/elastic-agent/pkg/agent/application/paths"
+	"github.com/elastic/beats/v7/x-pack/elastic-agent/pkg/agent/control"
+	"github.com/elastic/beats/v7/x-pack/elastic-agent/pkg/agent/control/client"
+	"github.com/elastic/beats/v7/x-pack/elastic-agent/pkg/agent/errors"
+	"github.com/elastic/beats/v7/x-pack/elastic-agent/pkg/core/logger"
+)
+
+const (
+	watcherSubcommand  = "watch"
+	maxRestartCount    = 5
+	restartBackoffInit = 5 * time.Second
+	restartBackoffMax  = 90 * time.Second
+)
+
+// Rollback rollbacks to previous version which was functioning before upgrade.
+func Rollback(ctx context.Context, prevHash, currentHash string) error {
+	// change symlink
+	if err := ChangeSymlink(ctx, prevHash); err != nil {
+		return err
+	}
+
+	// revert active commit
+	if err := UpdateActiveCommit(prevHash); err != nil {
+		return err
```

Comment:
I think you need to rollback the symlink on failure.

Reply:
i was thinking about this in a way, that reseting symlink is most important thing on rollback, if active commit, or cleanup is not changed it should not prevent running older (correct) version instead of failed new one. 
If restart fails correct agent should be started after Service Manager restarts agent either on machine restart or for whatever reason.